File: java/base/src/main/vsdk/toolkit/common/linealAlgebra/Vector2Df.java
package vsdk.toolkit.common.linealAlgebra;

import java.io.Serial;
import vsdk.toolkit.common.FundamentalEntity;
import vsdk.toolkit.common.VSDK;

public final class Vector2Df extends FundamentalEntity {
    @Serial
    private static final long serialVersionUID = 20260514L;

    private final float x;
    private final float y;

    public Vector2Df() {
        this(0f, 0f);
    }

    public Vector2Df(float x, float y) {
        this.x = x;
        this.y = y;
    }

    public Vector2Df(Vector2Df other) {
        this(other.x, other.y);
    }

    public Vector2Df(Vector2Dd other) {
        this((float) other.x(), (float) other.y());
    }

    public Vector2Df multiply(float a) {
        return new Vector2Df(a * x, a * y);
    }

    public float length() {
        return (float) Math.sqrt(x * x + y * y);
    }

    public Vector2Df add(Vector2Df b) {
        return new Vector2Df(x + b.x, y + b.y);
    }

    public Vector2Df withX(float nx) { return new Vector2Df(nx, y); }
    public Vector2Df withY(float ny) { return new Vector2Df(x, ny); }
    public float x() { return x; }
    public float y() { return y; }

    public boolean epsilonEquals(Vector2Df other) {
        return epsilonEquals(other, (float) VSDK.EPSILON);
    }

    public boolean epsilonEquals(Vector2Df other, float epsilon) {
        if ( other == null ) return false;
        if ( epsilon < 0f ) throw new IllegalArgumentException("epsilon must be >= 0");
        return Math.abs(x - other.x) <= epsilon && Math.abs(y - other.y) <= epsilon;
    }
}

File: typescript/base/src/main/vsdk/toolkit/common/linealAlgebra/Vector2Df.ts
import { FundamentalEntity } from "../FundamentalEntity.js";
import { VSDK } from "../VSDK.js";
import { Vector2Dd } from "./Vector2Dd.js";
const f = Math.fround;
export class Vector2Df extends FundamentalEntity {
  private readonly xv:number; private readonly yv:number;
  public constructor(); public constructor(x:number,y:number); public constructor(other:Vector2Df|Vector2Dd);
  public constructor(a:number|Vector2Df|Vector2Dd=0,b=0){super();if(typeof a==="number"){this.xv=f(a);this.yv=f(b);}else if(a instanceof Vector2Df){this.xv=a.x();this.yv=a.y();}else{this.xv=f(a.x);this.yv=f(a.y);}}
  public multiply(a:number):Vector2Df{return new Vector2Df(f(a*this.xv),f(a*this.yv));} public length():number{return f(Math.sqrt(f(this.xv*this.xv+this.yv*this.yv)));}
  public add(b:Vector2Df):Vector2Df{return new Vector2Df(f(this.xv+b.xv),f(this.yv+b.yv));} public withX(x:number):Vector2Df{return new Vector2Df(x,this.yv);} public withY(y:number):Vector2Df{return new Vector2Df(this.xv,y);} public x():number{return this.xv;} public y():number{return this.yv;}
  public epsilonEquals(o:Vector2Df|null,e=f(VSDK.EPSILON)):boolean{if(o===null)return false;if(e<0)throw new RangeError("epsilon must be >= 0");return Math.abs(this.xv-o.xv)<=e&&Math.abs(this.yv-o.yv)<=e;}
}
